source/SoundCore.cpp

```diff
@@ -182,39 +280,20 @@ HRESULT DSGetLock(LPDIRECTSOUNDBUFFER pVoice, DWORD dwOffset, DWORD dwBytes,
 
 //-----------------------------------------------------------------------------
 
+SoundBufferBase::CreateSoundBufferFunc SoundBufferBase::Create = NULL;
+
 HRESULT DSGetSoundBuffer(VOICE* pVoice, DWORD dwFlags, DWORD dwBufferSize, DWORD nSampleRate, int nChannels, const char* pszDevName)
 {
-	if (!g_lpDS)
-		return E_FAIL;
-
 	pVoice->name = pszDevName;
 
-	WAVEFORMATEX wavfmt;
-	DSBUFFERDESC dsbdesc;
-
-	wavfmt.wFormatTag = WAVE_FORMAT_PCM;
-	wavfmt.nChannels = nChannels;
-	wavfmt.nSamplesPerSec = nSampleRate;
-	wavfmt.wBitsPerSample = 16;
-	wavfmt.nBlockAlign = wavfmt.nChannels==1 ? 2 : 4;
-	wavfmt.nAvgBytesPerSec = wavfmt.nBlockAlign * wavfmt.nSamplesPerSec;
-
-	memset (&dsbdesc, 0, sizeof (dsbdesc));
-	dsbdesc.dwSize = sizeof (dsbdesc);
-	dsbdesc.dwBufferBytes = dwBufferSize;
-	dsbdesc.lpwfxFormat = &wavfmt;
-	dsbdesc.dwFlags = dwFlags | DSBCAPS_GETCURRENTPOSITION2 | DSBCAPS_STICKYFOCUS;
-	dsbdesc.szName = pszDevName;
-
-	// Are buffers released when g_lpDS OR pVoice->lpDSBvoice is released?
-	// . From DirectX doc:
-	//   "Buffer objects are owned by the device object that created them. When the
-	//    device object is released, all buffers created by that object are also released..."
-	HRESULT hr = g_lpDS->CreateSoundBuffer(&dsbdesc, &pVoice->lpDSBvoice, NULL);
-	if(FAILED(hr))
+	SoundBufferBase* soundBuffer = SoundBufferBase::Create();
```

Comment:
I admit I have not debugged the code, but how is this freed? Would it no tbe much easier to follow if the pointer was always wrppaed in a `shared_ptr`, here and when it is stored inside `VOICE`?